### DISCUSSION TITLE
SMS_AURA_UPDATE and HasCastUnit

### DISCUSSION THREAD

**Discussion by Golrag** [TrinityCore/TrinityCore#25986] [Q&A] [answered]
I got a sniff from back in legion (7.3.5) where a spell is cast by a gameobject yet HasCastUnit is set to true but there is no CastUnit. This spell is cast by The Alliance Flag in warsong gulch, with no CasterUnit.
```
ServerToClient: SMSG_AURA_UPDATE (0x2C22) Length: 49 ConnIdx: 0 Time: 03/16/2018
UpdateAll: False
AurasCount: 1
[0] Slot: 64
[0] HasAura: True
[0] CastID: Full: CASTGUID Map: Warsong Gulch Low: zzzzzzzz
[0] SpellID: 156621 (Alliance Flag)
[0] SpellXSpellVisualID: 66658
[0] Flags: 2 (Positive)
[0] ActiveFlags: 15
[0] CastLevel: 0
[0] Applications: 0
[0] HasCastUnit: True
[0] HasDuration: False
[0] HasRemaining: False
[0] HasTimeMod: False
[0] PointsCount: 0
[0] EstimatedPoints: 0
[0] HasSandboxScaling: False
[0] CastUnit: Full: 0x0 <<<< This is empty like that in the sniff.
UnitGUID: Full: PLAYERGUID
```

So I wonder why this field is set. Doubt its wrong. But it might just be `HasCaster`?

This is what I kinda want to achieve:
- Script the aura so that when it gets removed I can access the Caster GameObjects Template
- From the template get the entry of the gameobject that must be spawned
- Spawn the gameobject

This could clean up some code in BattlegroundWS

**Comment by Shauren** [comment, marked as answer]
Needs gameobject casting from 3.3.5, then you can implement

**Reply by Golrag** [reply]
So GameObject casting has been ported and I stumbled across this piece of code: https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Spells/Auras/SpellAuras.cpp#L395-L401

which makes it impossible to get the gameobject right now with this piece of code

```cpp
       // on aura remove
        if (GameObject* gameObjectCaster = player->GetMap()->GetGameObject(GetCasterGUID()))
            if (gameObjectCaster->GetGoType() == GAMEOBJECT_TYPE_NEW_FLAG)
                if (GameObject* droppedFlag = gameObjectCaster->SummonGameObject(gameObjectCaster->GetGOInfo()->newflag.FlagDrop, player->GetPosition(), QuaternionData::fromEulerAnglesZYX(player->GetOrientation(), 0.f, 0.f), gameObjectCaster->GetGOInfo()->newflag.ExpireDuration, GO_SUMMON_TIMED_DESPAWN))
                    droppedFlag->SetOwnerGUID(gameObjectCaster->GetGUID());
```

So this leaves me wondering how I should approach this right now. Should we let gameobjects be able to be caster from auras (and hide them in packet?) or should i go with spell/aura scripts and implement it in a less generic way?

**Reply by Shauren** [reply]
We should let auras have gameobject caster and not send it in packets
